tutorials/elixir-phoenix-on-kubernetes-google-container-engine/Dockerfile

```diff
@@ -0,0 +1,118 @@
+# It includes two stages and requires Docker 17.05 or later.
+
+# To adapt this Dockerfile to your app, set the APP_NAME and PHOENIX_SUBDIR
+# build arguments, as outlined below.
+
+
+################################################################################
+# BUILD STAGE
+#
+# Builds the app, does a brunch build to build the assets, and creates a
+# release. This stage uses the official Alpine Linux - Elixir base image, and
+# installs a few additional build tools such as Node.js.
+
+FROM elixir:alpine
+
+
+##############################################################################
+## Build arguments. Modify these to adapt this Dockerfile to your app.      ##
+## Alternatively, you may specify --build-arg when running `docker build`.  ##
+
+## The name of your Phoenix application.                                    ##
+ARG APP_NAME=hello
+
+## The subdirectory of the Phoenix application within the toplevel project. ##
+## If your project is an umbrella project, this should be the relative path ##
+## to the Phoenix application, e.g. "apps/hello_web". If this is a simple   ##
+## project, this should be "."                                              ##
+ARG PHOENIX_SUBDIR=.
+
+## End build arguments.                                                     ##
+##############################################################################
+
+
+# Set up build environment.
+ENV MIX_ENV=prod \
+    REPLACE_OS_VARS=true \
+    TERM=xterm
+
+# We will build the app in this directory.
+WORKDIR /opt/app
+
+# Install build tools needed in addition to Elixir:
+# NodeJS is used for brunch builds of Phoenix assets.
+# Hex and Rebar are needed to get and build dependencies.
+RUN apk update \
+    && apk --no-cache --update add nodejs nodejs-npm \
+    && mix local.rebar --force \
+    && mix local.hex --force
+
+# Copy the application files into /opt/app.
+COPY . .
+
+# Build the application.
+RUN mix do deps.get, deps.compile, compile
+
+# Build assets by running a brunch build and phoenix digest.
+# If you are using a different mechanism for asset builds, you may need to
+# alter these commands.
+RUN cd ${PHOENIX_SUBDIR}/assets \
+    && npm install \
+    && ./node_modules/brunch/bin/brunch build -p \
+    && cd .. \
+    && mix phx.digest
+
+# Create the release, and move the artifacts to well-known paths so the
+# runtime image doesn't need to know the app name. Specifically, the binary
+# is renamed to "start_server", and the entire release is moved into
+# /opt/release.
+RUN mix release --env=prod --verbose \
+    && mv _build/prod/rel/${APP_NAME} /opt/release \
+    && mv /opt/release/bin/${APP_NAME} /opt/release/bin/start_server
+
+
+################################################################################
+# RUNTIME STAGE
+#
+# Creates the actual runtime image. This is based on a the Alpine Linux base
+# image, with only the minimum dependencies for running ERTS.
+
+FROM alpine:latest
```

Comment:
Nice use of the multi-stage docker syntax!  We should add a note somewhere that calls out the min version of docker required to run this.

Reply:
The "before you begin" section of the tutorial does say "Install Docker 17.05 or later." Or do you mean in the Dockerfile itself?

Reply:
I noticed that further down.  Disregard :)